GSoC 2018 cache project

javierron opened the pull request.

javierron committed:
ceph mempool imported.Signed-off-by: Javier Ron <javierron@outlook.com>

javierron committed:
cache: working local cahce impl, eviction policies availabe: LRU & ARC

Signed-off-by: Javier Ron <javierron@outlook.com>

javierron committed:
build: update tested os versions

Signed-off-by: Javier Ron <javierron@outlook.com>

javierron committed:
Modified bench2 to measure read iops latency and throughput.
Reads are generated by a normal distribution

Signed-off-by: Javier Ron <javierron@outlook.com>

modified run.sh

javierron committed:
Inserted macros enable/disable cache and stats

Signed-off-by: Javier Ron <javierron@outlook.com>

javierron committed:
Create options rst file

Update index.rst

updated options rst

added options doc to toctree

Signed-off-by: Javier Ron <javierron@outlook.com>

dotnwat commented on src/include/CMakeLists.txt:
```diff
@@ -6,6 +6,10 @@ install(FILES
     zlog/stream.h
     zlog/options.h
     zlog/statistics.h
+    zlog/mempool/mempool.h
```
Do these headers need to be installed?

dotnwat replied:
They may need to be installed if the have things that installed headers depend on. If they are strictly used internally, they don't need to be installed.

dotnwat commented on src/include/CMakeLists.txt:
```diff
@@ -18,7 +22,17 @@ if(BUILD_CEPH_BACKEND)
+  zlog/eviction/arc.h)
+install(FILES
+  ${eviction_hdrs}
+  DESTINATION include/zlog/eviction
```
Same thing there. Need to be installed?

dotnwat commented on src/java/native/zlogjni.cc:
```diff
@@ -11,6 +11,8 @@ void Java_org_cruzdb_zlog_Log_disposeInternal(
   delete reinterpret_cast<zlog::Log*>(jhandle);
 }
 
+zlog::Options options;
```
This definitely looks like a nasty bug waiting to happen. Why move options out of the stack allocation?

dotnwat commented on src/monitoring/statistics.h:
```diff
@@ -100,7 +100,7 @@ inline void MeasureTime(Statistics* statistics, uint32_t histogram_type,
 }
 
-inline void RecordTick(Statistics* statistics, uint32_t ticker_type,
+inline void RecordTick(std::shared_ptr<Statistics> statistics, uint32_t ticker_type,
```
why is this changed to a shared pointer? this was copied straight from rocksdb, right?

dotnwat commented on CMakeLists.txt:
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 2.8.12)
 project(zlog)
 
+add_definitions(-DWITH_CACHE -DWITH_STATS)
```
This is close. I think we want to have a cmake option so we can turn this on and off. Something like

```
option(WITH_CACHING "add caching" OFF)
if(WITH_CACHING)
  add_definitions(-DWITH_CACHE)
endif()
```

dotnwat reviewed (commented):
Overall this looks good. I left detailed comments for pretty minor changes. I'd like to sort out the situation with the headers. I think some don't need to be installed, and can be moved out of the include directory. The only other thing that'd be nice (but not really a big deal) is a removing the whitespace changes.

javierron committed:
pull request fixes:
- fixed indentation to match project standard
- Make a copy of options struct on instancing log
- no longer installing libraries not used by client apps
- added cmake option WITH_CACHING

Signed-off-by: Javier Ron <javierron@outlook.com>

dotnwat commented:
WEEEEEEEEEEEEEEEeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee
